glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt

```diff
@@ -195,9 +196,15 @@ internal object Dispatchers {
         }
     }
 
+    // This job is used to make sure the API `CoroutineContext` does not cancel
+    // children jobs when exceptions are thrown in children coroutines.
+    private val supervisorJob = SupervisorJob()
+
     /**
      * A coroutine scope to make it easy to dispatch API calls off the main thread.
      * This needs to be a `var` so that our tests can override this.
      */
-    var API = WaitableCoroutineScope(CoroutineScope(newSingleThreadContext("GleanAPIPool")))
+    var API = WaitableCoroutineScope(CoroutineScope(
+        newSingleThreadContext("GleanAPIPool") + supervisorJob
```

Comment:
Uh, how does this work? Does whatever `newSingleThreadContext` returns somehow override the addition?

Reply:
Basically, [`CoroutineScope`](https://github.com/Kotlin/kotlinx.coroutines/blob/9be85a44f157cea6bf14b685a87c6709ebd9664b/kotlinx-coroutines-core/common/src/CoroutineScope.kt#L188) acquires a context in the constructor. [`newSingleThreadContext`](https://github.com/Kotlin/kotlinx.coroutines/blob/9be85a44f157cea6bf14b685a87c6709ebd9664b/kotlinx-coroutines-core/jvm/src/ThreadPoolDispatcher.kt#L30) returns a `CoroutineDispatcher` (that is, if you inspect the class hierarchy deep enough...) which is basically a `Context`. The `SupervisorJob` is itself a context (again, investigate deep enough and it boils down to a `CoroutineContext.Element`). The plus operator is defined for `CoroutineContext` + `CoroutineContext` [here](https://github.com/JetBrains/kotlin/blob/db91b520a5f5d056cb711c0531862536e7bd12c7/libraries/stdlib/src/kotlin/coroutines/CoroutineContext.kt#L30), creating a `CombinedContext` which takes the `childCancelled` from `SupervisorCoroutine`, which, in the end, is responsible for not making [things explode](https://github.com/Kotlin/kotlinx.coroutines/blob/9be85a44f157cea6bf14b685a87c6709ebd9664b/kotlinx-coroutines-core/common/src/Supervisor.kt#L57)

Reply:
That ... was a deep dive, thanks.